File: kotlin/src/main/kotlin/benchmarks/JsonGenerate.kt
package benchmarks

import Benchmark
import com.alibaba.fastjson2.JSONArray
import com.alibaba.fastjson2.JSONObject
import java.util.Locale

class JsonGenerate : Benchmark() {
    var n = configVal("coords")
    private lateinit var data: List<Map<String, Any>>
    lateinit var text: String
    private var resultVal: Long = 0

    override fun prepare() {
        data =
            List(n.toInt()) {
                mapOf(
                    "x" to "%.8f".format(Locale.US, Helper.nextFloat()).toDouble(),
                    "y" to "%.8f".format(Locale.US, Helper.nextFloat()).toDouble(),
                    "z" to "%.8f".format(Locale.US, Helper.nextFloat()).toDouble(),
                    "name" to "${"%.7f".format(Locale.US, Helper.nextFloat())} ${Helper.nextInt(10000)}",
                    "opts" to mapOf("1" to listOf(1, true)),
                )
            }
    }

    override fun run(iterationId: Int) {
        val jsonArray = JSONArray()
        for (coord in data) {
            jsonArray.add(coord)
        }

        val jsonObject = JSONObject()
        jsonObject.put("coordinates", jsonArray)
        jsonObject.put("info", "some info")

        text = jsonObject.toJSONString()
        if (text.startsWith("{\"coordinates\":")) resultVal += 1
    }

    override fun checksum(): UInt = resultVal.toUInt()

    override fun name(): String = "Json::Generate"
}

File: nim/src/benchmarks/jsongenerate.nim
import std/[strutils]
import jsony
import ../benchmark
import ../helper
import jsonbench_common

type
  JsonGenerate* = ref object of Benchmark
    n: int64
    data: JsonData
    resultJson: string
    resultVal: uint32

proc newJsonGenerate(): Benchmark =
  JsonGenerate()

method name(self: JsonGenerate): string = "JsonGenerate"

method prepare(self: JsonGenerate) =
  self.n = self.config_val("coords")
  self.resultVal = 0
  self.data = generateJsonData(self.n)

method run(self: JsonGenerate, iteration_id: int) =
  self.resultJson = self.data.toJson()
  if self.resultJson.startsWith("{\"coordinates\":"):
    self.resultVal = self.resultVal + 1

method checksum(self: JsonGenerate): uint32 =
  self.resultVal

proc getGeneratedJson*(self: JsonGenerate): string =
  self.resultJson

registerBenchmark("JsonGenerate", newJsonGenerate)